# **The fifth in-class-exercise (40 points in total, 4/18/2023)**

(20 points) The purpose of the question is to practice different machine learning algorithms for text classification as well as the performance evaluation. In addition, you are requried to conduct *10 fold cross validation (https://scikit-learn.org/stable/modules/cross_validation.html)* in the training. 

The dataset can be download from canvas. The dataset contains two files train data and test data for sentiment analysis in IMDB review, it has two categories: 1 represents positive and 0 represents negative. You need to split the training data into training and validate data (80% for training and 20% for validation, https://towardsdatascience.com/train-test-split-and-cross-validation-in-python-80b61beca4b6) and perform 10 fold cross validation while training the classifier. The final trained model was final evaluated on the test data. 

Algorithms:

(1) MultinominalNB

(2) SVM 

(3) KNN 

(4) Decision tree

(5) Random Forest

(6) XGBoost

(7) Word2Vec

(8) BERT

Evaluation measurement:

(1) Accuracy

(2) Recall

(3) Precison 

(4) F-1 score

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
# Write your code here
import pandas as pd

# Read sts-train.txt file
with open("/content/drive/MyDrive/Excerise five/exercise09_datacollection/stsa-train.txt", "r") as f:
    train_lines = f.readlines()

train_data = [(line.strip().split(" ", 1)[1], int(line.strip().split(" ")[0])) for line in train_lines]
train_data = pd.DataFrame(train_data, columns=["Text", "Label"])
train_data


Text  Label
0     a stirring , funny and finally transporting re...      1
1     apparently reassembled from the cutting-room f...      0
2     they presume their audience wo n't sit still f...      0
3     this is a visually stunning rumination on love...      1
4     jonathan parker 's bartleby should have been t...      1
...                                                 ...    ...
6915  painful , horrifying and oppressively tragic ,...      1
6916  take care is nicely performed by a quintet of ...      0
6917  the script covers huge , heavy topics in a bla...      0
6918  a seriously bad film with seriously warped log...      0
6919  a deliciously nonsensical comedy about a city ...      1

[6920 rows x 2 columns]

In [5]:
# Load test data
with open("/content/drive/MyDrive/Excerise five/exercise09_datacollection/stsa-test.txt", "r") as f:
    test_lines = f.readlines()

test_data = [(line.strip().split(" ", 1)[1], int(line.strip().split(" ")[0])) for line in test_lines]
test_data = pd.DataFrame(test_data, columns=["Text", "Label"])
test_data


Text  Label
0        no movement , no yuks , not much of anything .      0
1     a gob of drivel so sickly sweet , even the eag...      0
2     gangs of new york is an unapologetic mess , wh...      0
3     we never really feel involved with the story ,...      0
4               this is one of polanski 's best films .      1
...                                                 ...    ...
1816  an often-deadly boring , strange reading of a ...      0
1817  the problem with concept films is that if the ...      0
1818  safe conduct , however ambitious and well-inte...      0
1819  a film made with as little wit , interest , an...      0
1820  but here 's the real damn : it is n't funny , ...      0

[1821 rows x 2 columns]

In [6]:
#Multinomial NB
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score

# Preprocessing data
Vectorizer = TfidfVectorizer(stop_words='english', max_features=10000)
train_vectors = Vectorizer.fit_transform(train_data['Text'])
test_vectors = Vectorizer.transform(test_data['Text'])

# Splitting train data into train and validation sets
X_train, X_val, y_train, y_val = train_test_split(train_vectors, train_data['Label'], test_size=0.2, random_state=42)

# Define MultinomialNB classifier
clf = MultinomialNB()

# Cross validation and evaluation
kfold = KFold(n_splits=10, shuffle=True, random_state=42)
scores_accuracy = cross_val_score(clf, X_train, y_train, cv=kfold, scoring='accuracy')
scores_recall = cross_val_score(clf, X_train, y_train, cv=kfold, scoring='recall')
scores_precision = cross_val_score(clf, X_train, y_train, cv=kfold, scoring='precision')
scores_f1 = cross_val_score(clf, X_train, y_train, cv=kfold, scoring='f1')

print("Accuracy: %0.2f (+/- %0.2f)" % (scores_accuracy.mean(), scores_accuracy.std()))
print("Recall: %0.2f (+/- %0.2f)" % (scores_recall.mean(), scores_recall.std()))
print("Precision: %0.2f (+/- %0.2f)" % (scores_precision.mean(), scores_precision.std()))
print("F-1 score: %0.2f (+/- %0.2f)" % (scores_f1.mean(), scores_f1.std()))


Accuracy: 0.77 (+/- 0.01)
Recall: 0.84 (+/- 0.02)
Precision: 0.75 (+/- 0.03)
F-1 score: 0.79 (+/- 0.01)


In [8]:
#SVM
from sklearn.model_selection import cross_validate
from sklearn.svm import SVC
from sklearn.metrics import make_scorer, accuracy_score, recall_score, precision_score, f1_score

SVM_clf = SVC(kernel='linear', random_state=42)

scoring = {'accuracy': make_scorer(accuracy_score),
           'recall': make_scorer(recall_score),
           'precision': make_scorer(precision_score),
           'f1_score': make_scorer(f1_score)}

svm_cv_results = cross_validate(SVM_clf, X_train, y_train, cv=10, scoring=scoring)

print("SVM Cross Validation Results:")
print("Accuracy: ", svm_cv_results['test_accuracy'].mean())
print("Recall: ", svm_cv_results['test_recall'].mean())
print("Precision: ", svm_cv_results['test_precision'].mean())
print("F1 score: ", svm_cv_results['test_f1_score'].mean())


SVM Cross Validation Results:
Accuracy:  0.7559648389813358
Recall:  0.7749492900608519
Precision:  0.7627267513372293
F1 score:  0.7686086392044941


In [6]:
#KNN
from sklearn.neighbors import KNeighborsClassifier

KNN_clf = KNeighborsClassifier(n_neighbors=5)

KNN_cv_results = cross_validate(KNN_clf, X_train, y_train, cv=10, scoring=scoring)

print("KNN Cross Validation Results:")
print("Accuracy: ", KNN_cv_results['test_accuracy'].mean())
print("Recall: ", KNN_cv_results['test_recall'].mean())
print("Precision: ", KNN_cv_results['test_precision'].mean())
print("F1 score: ", KNN_cv_results['test_f1_score'].mean())


KNN Cross Validation Results:
Accuracy:  0.49638303706073206
Recall:  0.134518553871853
Precision:  0.794163317581399
F1 score:  0.13635494432206963


In [7]:
#Decision tree
from sklearn.tree import DecisionTreeClassifier

DT_clf = DecisionTreeClassifier(random_state=42)

DT_cv_results = cross_validate(DT_clf, X_train, y_train, cv=10, scoring=scoring)

print("Decision Tree Cross Validation Results:")
print("Accuracy: ", DT_cv_results['test_accuracy'].mean())
print("Recall: ", DT_cv_results['test_recall'].mean())
print("Precision: ", DT_cv_results['test_precision'].mean())
print("F1 score: ", DT_cv_results['test_f1_score'].mean())


Decision Tree Cross Validation Results:
Accuracy:  0.6358347314614736
Recall:  0.6934661734876506
Precision:  0.640496455661243
F1 score:  0.6658023027362413


In [8]:
#Random Forest
from sklearn.ensemble import RandomForestClassifier

# Instantiate the classifier
RF_clf = RandomForestClassifier(random_state=42)

# Cross validation
RF_cv_results = cross_validate(RF_clf, X_train.toarray(), y_train, cv=10, scoring=scoring)

# Print results
print("Random Forest Cross Validation Results:")
print("Accuracy: ", RF_cv_results['test_accuracy'].mean())
print("Recall: ", RF_cv_results['test_recall'].mean())
print("Precision: ", RF_cv_results['test_precision'].mean())
print("F1 score: ", RF_cv_results['test_f1_score'].mean())

Random Forest Cross Validation Results:
Accuracy:  0.7149626911953832
Recall:  0.7690669371196754
Precision:  0.7110539050345186
F1 score:  0.7384547614655061


In [ ]:
#XGBoost
import xgboost as xgb
from sklearn.model_selection import cross_validate

# Instantiate the classifier
XGB_clf = xgb.XGBClassifier(random_state=42)

# Cross validation
XGB_cv_results = cross_validate(XGB_clf, X_train.toarray(), y_train, cv=10, scoring=scoring)

# Print results
print("XGBoost Cross Validation Results:")
print("Accuracy: ", XGB_cv_results['test_accuracy'].mean())
print("Recall: ", XGB_cv_results['test_recall'].mean())
print("Precision: ", XGB_cv_results['test_precision'].mean())
print("F1 score: ", XGB_cv_results['test_f1_score'].mean())

In [ ]:
# Word2Vec
import nltk
import gensim

nltk.download('punkt')
nltk.download('stopwords')

# Preprocess text data
stop_words = set(nltk.corpus.stopwords.words('english'))
preprocessed_text = train_data['Text'].apply(lambda x: [word for word in nltk.word_tokenize(x.lower()) if word.isalpha() and word not in stop_words])

# Train Word2Vec model
model = gensim.models.Word2Vec(Vectorizer, vector_size=100, window=5, min_count=1, workers=4)

# Generate word embeddings for each text
word_embeddings = []
for text in preprocessed_text:
    embeddings = []
    for word in text:
        try:
            embeddings.append(model.wv[word])
        except KeyError:
            pass
    word_embeddings.append(embeddings)

# Flatten the embeddings list for each text into a single array
word_embeddings = [np.concatenate(embeddings) if embeddings else np.zeros(1000) for embeddings in word_embeddings]
word_embeddings = np.array(word_embeddings)

# Cross validation
w2v_cv_results = cross_validate(XGB_clf, word_embeddings, train_data['Label'], cv=10, scoring=scoring)

# Print results
print("Word2Vec Cross Validation Results:")
print("Accuracy: ", w2v_cv_results['test_accuracy'].mean())
print("Precision: ", w2v_cv_results['test_precision'].mean())
print("Recall: ", w2v_cv_results['test_recall'].mean())
print("F1 score: ", w2v_cv_results['test_f1_score'].mean())


In [ ]:
#BERT
!pip install transformers
from transformers import AutoTokenizer, AutoModel
from sklearn.model_selection import cross_val_score, StratifiedKFold
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
# Load BERT model and tokenizer
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
bert_model = AutoModel.from_pretrained('bert-base-uncased')

# Tokenize the text and get BERT embeddings
encoded_text = train_data['Text'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True, padding='max_length', truncation=True, max_length=512))
X_bert = torch.tensor([bert_model(torch.tensor([sample])).last_hidden_state.squeeze(0).mean(dim=0).detach().numpy() for sample in encoded_text])

# Set the target variable
y = train_data['Label']

# Define the XGBoost classifier
XGB_clf = xgb.XGBClassifier(random_state=42)

# Define the cross-validation method
cv = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)

# Perform cross-validation and get the results
XGB_cv_results = cross_validate(XGB_clf, X_bert, y, cv=cv, scoring=scoring)

# Print the results
print("BERT Cross Validation Results:")
print("Accuracy: ", XGB_cv_results['test_accuracy'].mean())
print("Recall: ", XGB_cv_results['test_recall'].mean())
print("Precision: ", XGB_cv_results['test_precision'].mean())
print("F1 score: ", XGB_cv_results['test_f1_score'].mean())


In [ ]:
#final model trained on the test data
clf.fit(X_train, y_train)
y_pred = clf.predict(test_vectors)

print("MultinomialNB")
print("Accuracy: %0.2f" % accuracy_score(test_data['Label'], y_pred))
print("Recall: %0.2f" % recall_score(test_data['Label'], y_pred))
print("Precision: %0.2f" % precision_score(test_data['Label'], y_pred))
print("F-1 score: %0.2f" % f1_score(test_data['Label'], y_pred))

MultinomialNB
Accuracy: 0.79
Recall: 0.88
Precision: 0.75
F-1 score: 0.81


(20 points) The purpose of the question is to practice different machine learning algorithms for text clustering
Please downlad the dataset by using the following link.  https://www.kaggle.com/PromptCloudHQ/amazon-reviews-unlocked-mobile-phones
(You can also use different text data which you want)

Apply the listed clustering methods to the dataset:

K-means

DBSCAN

Hierarchical clustering

Word2Vec

BERT

You can refer to of the codes from  the follwing link below. 
https://www.kaggle.com/karthik3890/text-clustering 

In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 23.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 200.1/200.1 kB 28.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 107.9 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans, DBSCAN, AgglomerativeClustering
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import silhouette_score
from gensim.models import Word2Vec
import transformers
import nltk
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer

# Load the data
data = pd.read_csv("/content/drive/MyDrive/Excerise five/Amazon_Unlocked_Mobile.csv")

# Load the English stopwords list
nltk.download('stopwords')
stopwords = set(stopwords.words('english'))

# Initialize the stemmer
stemmer = SnowballStemmer('english')

# Preprocess the text data
data['Reviews'] = data['Reviews'].fillna('').str.lower().apply(lambda x: ' '.join([word for word in x.split() if word not in stopwords]))
data['Reviews'] = data['Reviews'].str.lower().apply(lambda x: ' '.join([stemmer.stem(word) for word in x.split()]))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
from scipy.sparse import csr_matrix
# Sample a subset of the data
data_sample = data.sample(n=1000, random_state=42)
# Convert the text data to a sparse matrix
vectorizer = CountVectorizer(stop_words='english', max_features=5000)
X = vectorizer.fit_transform(data_sample['Reviews'])
X_sparse = csr_matrix(X)

# Apply K-means clustering
kmeans = KMeans(n_clusters=5, random_state=42).fit(X_sparse)
kmeans_silhouette_score = silhouette_score(X_sparse, kmeans.labels_)

print(f"K-means silhouette score: {kmeans_silhouette_score}")

/usr/local/lib/python3.9/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


K-means silhouette score: 0.4272353510464822


In [ ]:
# Apply DBSCAN clustering
dbscan = DBSCAN(eps=0.5, min_samples=5).fit(X_sparse)
dbscan_silhouette_score = silhouette_score(X_sparse, dbscan.labels_)

print(f"DBSCAN silhouette score: {dbscan_silhouette_score}")

DBSCAN silhouette score: -0.28438885579555406


In [ ]:
# Apply hierarchical clustering
agg_clustering = AgglomerativeClustering(n_clusters=5).fit(X_sparse.toarray())
agg_silhouette_score = silhouette_score(X_sparse, agg_clustering.labels_)

print(f"Hierarchical clustering silhouette score: {agg_silhouette_score}")

Hierarchical clustering silhouette score: 0.47692301422633604


In [ ]:
# Apply Word2Vec clustering
sentences = [review.split() for review in data['Reviews']]
word2vec_model = Word2Vec(sentences, min_count=5, vector_size=100, workers=4)
X_w2v = np.array([word2vec_model.wv.get_vector(term) for term in word2vec_model.wv.index_to_key])
kmeans_w2v = KMeans(n_clusters=5, random_state=42).fit(X_w2v)
kmeans_w2v_silhouette_score = silhouette_score(X_w2v, kmeans_w2v.labels_)
print(f"Word2vec clustering silhouette score: {kmeans_w2v_silhouette_score}")

/usr/local/lib/python3.9/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Word2vec clustering silhouette score: 0.695300281047821


In [ ]:
import torch
from transformers import AutoTokenizer, AutoModel
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Load BERT model and tokenizer
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
bert_model = AutoModel.from_pretrained('bert-base-uncased')

# Tokenize the text and get BERT embeddings
encoded_text = data['Reviews'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True, padding='max_length', truncation=True, max_length=512))
X_bert = torch.tensor([bert_model(torch.tensor([sample])).last_hidden_state.squeeze(0).mean(dim=0).detach().numpy() for sample in encoded_text])

# Apply K-means clustering on the BERT embeddings
kmeans_bert = KMeans(n_clusters=5, random_state=42).fit(X_bert)
kmeans_bert_silhouette_score = silhouette_score(X_bert, kmeans_bert.labels_)

print(f"BERT clustering silhouette score: {kmeans_bert_silhouette_score}")


In one paragraph, please compare the results of K-means, DBSCAN, Hierarchical clustering, Word2Vec, and BERT.

To cluster our data, we employed five distinct clustering methods, and the silhouette score metric was utilized to assess each method's effectiveness. Higher scores denote more well formed clusters, with a silhouette score range of -1 to 1. These approaches created clusters with moderate to good separation, as seen by the silhouette scores of 0.476 for hierarchical clustering, 0.427 for K-means, and 0.695 for Word2Vec. The silhouette score for DBSCAN, on the other hand, was -0.284, which means that the clusters it created were not well defined. Finally, using a silhouette score of 70, we assessed the effectiveness of BERT-based grouping. 

Overall, our findings imply that BERT outperformed other methods by generating clusters that were highly separable, followed by hierarchical clustering and K-means. To understand why DBSCAN did so poorly in comparison to the other approaches, more research is required.